Configurations.md

```diff
@@ -2379,14 +2379,48 @@ Don't reformat out of line modules
 
 ## `single_line_if_else_max_width`
 
-Maximum line length for single line if-else expressions. A value of `0` (zero) results in if-else expressions always being broken into multiple lines. Note this occurs when `use_small_heuristics` is set to `Off`.
+Maximum line length for single line if-else expressions. A value of `0` (zero) results in if-else expressions always being broken into multiple lines. Note this occurs when `use_small_heuristics` is set to `Off`. If one of the if-else blocks contain a statement, a comment, an attribute, or is empty, the if-else expression will be broken into multiple lines. When using `version = "Two"` the if-else expression at the end of a block will be formatted on a single line. The last if-else expression in the last example snippet of this section would be formatted on a single line with `version = "Two"`.
```

Comment:
> When using `version = "Two"` the if-else expression at the end of a block will be formatted on a single line

Are you saying that this would only apply to the last if-else expression?  I don't think that's correct, but I could be wrong.

Reply:
Thanks for having a look! I agree that this behavior sounds strange and maybe my research and testing has come up short here, or maybe the proposed documentation is not clear enough. Let me try to explain what I see happening. When we try to use `single_line_if_else_max_width` in expr.rs `self.allow_single_line` needs to be `true` (see [here](https://github.com/truls-p/rustfmt/blob/improve-docs-if-else-max/src/expr.rs#L922)), which only happens when we have an `if` and expression type is == `ExprType::SubExpression` (see [here](https://github.com/truls-p/rustfmt/blob/improve-docs-if-else-max/src/expr.rs#L649)). Now, the `version = "Two"` is relevant [here](https://github.com/truls-p/rustfmt/blob/improve-docs-if-else-max/src/stmt.rs#L78) where we set `expr_type` to `ExprType::SubExpression`. The two tests I added, 70.rs and 70_version_two.rs tries to test this specific behavior.

Here is also another example to play around with. If you run rustfmt --check on this you will see we get a diff for both functions. However, if we change version to "Two" we only get a diff for the last `if` in `foo`.
Config options:
```
single_line_if_else_max_width = 80
version = "One"
```
main.rs:
```rust
fn bar() {}
fn foo() -> usize {
    let some_long_name = true;
    if some_long_name && some_long_name && some_long_name { bar() } else { bar() }
    return 1
}

fn baz() {
    let some_long_name = true;
    if some_long_name && some_long_name && some_long_name { bar() } else { bar() }
}

fn main() {
    let i = foo();
    baz();
    println!("{}", i);
}

```

Reply:
Here are some other examples to consider:

What about something like this where the `if-else` is on the right hand side of an assignment expression?

```rust
fn foo() -> usize {
    let x = if some_long_name && some_long_name && some_long_name { bar() } else { bar() };
    let some_long_name = true;
    x
}
```

Also what about conditionally passing an argument to a function call or returning a value from a closure?
```rust
fn main() {
    let x = 0;
    f(if x == 0 { 0 } else { 10 })
    g(|x| if x == 0 { 0 } else { 10 })
}
```

Again, really appreciate your work on this, and helping us to nail down the description for this option. I think a little rewording is still needed, and maybe it's worth investigating how we set the `expr_type` argument. For example, when we call `rewrite` on a top level `ast::Expr` node we always pass `expr_type` as `ExprType::SubExpression`. See [`impl Rewrite for ast::Expr`](https://github.com/rust-lang/rustfmt/blob/38659ec6ad5f341cf8eb3139725bf695872c6de7/src/expr.rs#L39-L43). The only place in the code base I could find where we don't hard code the `expr_type` argument to `format_expr` is when [it gets called from `format_stmt`](https://github.com/rust-lang/rustfmt/blob/38659ec6ad5f341cf8eb3139725bf695872c6de7/src/stmt.rs#L111).

We happen to define our own borrowed version of a `ast::Stmt`:

https://github.com/rust-lang/rustfmt/blob/38659ec6ad5f341cf8eb3139725bf695872c6de7/src/stmt.rs#L13-L16

And the rewrite method for that does follow the behavior that you outlined above where the last expression is set to a `ExprType::SubExpression` if `Version=Two` is also set

https://github.com/rust-lang/rustfmt/blob/38659ec6ad5f341cf8eb3139725bf695872c6de7/src/stmt.rs#L76-L85

So maybe that's what explains the last expression behavior that you documented.


I also came across [this interesting comment](https://github.com/rust-lang/rustfmt/blob/38659ec6ad5f341cf8eb3139725bf695872c6de7/src/expr.rs#L1038-L1041) in the code base that would be nice to have a test for.

> If one of the if-else blocks contain a statement, a comment, an attribute, or is empty, the if-else expression will be broken into multiple lines. 

It would also be great to have tests for each of these cases that you outlined!

Reply:
Thanks for the detailed feedback so far!
These are good examples and I have added them to the documentation in the example section and mentioned them explicitly (the let, function call, and closure) in the description of the config option. I have also reworded the part about if-else expressions at the end of blocks. More on this below.

I have also added some more test cases for the nested if-else case (note how the inner if-else is on a single line when using `Version = "Two"` because it is the last expression in the block) and the comment about the if-else blocks needing to  be simple.

Regarding the `Version = "Two"` and how the `expr_type` is set I believe I now have a slightly better understanding about how this all works. My understanding is that as we walk the nodes everything starts out as a `Stmt`. This is why, for `Version = "One"`, when we look at an if-else at the end of a block we hit the following block of code and `expr_type` is set to `ExprType::Statement`. 

https://github.com/rust-lang/rustfmt/blob/38659ec6ad5f341cf8eb3139725bf695872c6de7/src/stmt.rs#L76-L85

However, in the let, function call, and closure examples we set the `expr_type` for the if-else much further down the call stack. As an example, for the if-else inside the function call we call we hit `rewrite_cond` where the `expr_type` is set to `ExprType::SubExpression` by default.
https://github.com/rust-lang/rustfmt/blob/38659ec6ad5f341cf8eb3139725bf695872c6de7/src/expr.rs#L610-L610

Then, crucially, `expr_type == ExprType::SubExpression` evaluates to true in `fn to_control_flow`.
https://github.com/rust-lang/rustfmt/blob/38659ec6ad5f341cf8eb3139725bf695872c6de7/src/expr.rs#L645-L658

On the other hand, for the if-else at the end of a block we go straight from `impl<'a> Rewrite for Stmt<'a>` to `fn format_stmt` to `fn format_expr`, and finally to `fn control_flow`, where the `expr_type` is either `ExprType::Statement` or `ExprType::SubExpression` (set in `impl<'a> Rewrite for Stmt<'a>`) depending on whether we use `Version = "One"` or `Version = "Two"`.

I also dug up the original PR for the `impl<'a> Rewrite for Stmt<'a>`:
https://github.com/rust-lang/rustfmt/pull/3631
Note the tests added for #3614: https://github.com/rust-lang/rustfmt/pull/3631/files#diff-c331b0359e60165a708105a49c54a82dc01be48cd219a4c83f91ba0ba961dcb2

Please let me know your thoughts!